vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -96,99 +98,108 @@ WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString, vector<C
 
 WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
 {
-  Synchronize();
+  Close(CloseCode::GoingAway, "Disposed");
+
+  m_closePerformed.Wait();
 }
 
 #pragma region Private members
 
 IAsyncAction WinRTWebSocketResource::PerformConnect()
 {
+  auto self = shared_from_this();
   try
   {
     co_await resume_background();
 
-    co_await m_socket.ConnectAsync(m_uri);
+    co_await self->m_socket.ConnectAsync(self->m_uri);
 
-    m_readyState = ReadyState::Open;
-    if (m_connectHandler)
+    self->m_readyState = ReadyState::Open;
+    if (self->m_connectHandler)
     {
-      m_connectHandler();
+      self->m_connectHandler();
     }
   }
   catch (hresult_error const& e)
   {
-    m_readyState = ReadyState::Closed;
-    if (m_errorHandler)
+    self->m_readyState = ReadyState::Closed;
+    if (self->m_errorHandler)
     {
-      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
     }
   }
 
-  SetEvent(m_connectPerformed.get());
-  m_connectPerformedPromise.set_value();
-  m_connectRequested = false;
+  SetEvent(self->m_connectPerformed.get());
+  self->m_connectPerformedPromise.set_value();
+  self->m_connectRequested = false;
 }
 
 fire_and_forget WinRTWebSocketResource::PerformPing()
 {
+  auto self = shared_from_this();
   try
   {
     co_await resume_background();
 
-    co_await resume_on_signal(m_connectPerformed.get());
+    co_await resume_on_signal(self->m_connectPerformed.get());
 
-    if (m_readyState != ReadyState::Open)
+    if (self->m_readyState != ReadyState::Open)
     {
+      self = nullptr;
       co_return;
     }
 
-    m_socket.Control().MessageType(SocketMessageType::Utf8);
+    self->m_socket.Control().MessageType(SocketMessageType::Utf8);
 
     string s{};
     winrt::array_view<const uint8_t> arr(
       CheckedReinterpretCast<const uint8_t*>(s.c_str()),
       CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
-    m_writer.WriteBytes(arr);
+    self->m_writer.WriteBytes(arr);
 
-    co_await m_writer.StoreAsync();
+    co_await self->m_writer.StoreAsync();
 
-    if (m_pingHandler)
+    if (self->m_pingHandler)
     {
-      m_pingHandler();
+      self->m_pingHandler();
     }
   }
   catch (hresult_error const& e)
   {
-    if (m_errorHandler)
+    if (self->m_errorHandler)
     {
-      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
     }
   }
 }
 
 fire_and_forget WinRTWebSocketResource::PerformWrite()
 {
-  if (m_writeQueue.empty())
+  auto self = shared_from_this();
+  if (self->m_writeQueue.empty())
   {
+    self = nullptr;
     co_return;
   }
 
+  self = shared_from_this();
```

Comment:
Why do we need this?

Reply:
We don't. Leftover from a `weak_ptr` prototype.
Removing.